File: zeta_amtct.py
import mpmath as mp
from statistics import NormalDist

def omega(sigma, t):
    return 1 - (2**(2-sigma))*mp.cos(t*mp.log(2))+4**(1-sigma)

def C(k0, k1, miu_n, sigma_n, k):
    if (k0 < k1):
        return 1
    return 1 - NormalDist().cdf((k - miu_n)/sigma_n)

def zeta_amtct(sigma, t, d):
    sigma = mp.mpf(sigma)
    t = mp.mpf(t)
    n = mp.pi * abs(t) + (1 + 2 * sigma)*mp.log(abs(t)) + 2*mp.log(mp.gamma(sigma)) - 2*mp.log(abs( 1 - 2 ** (1-sigma))) + 2*d*mp.log(10) + mp.log(2) - mp.log(mp.pi)
    n = n/(2*mp.log(3+mp.sqrt(8)))
    epsilon = 10 ** -d
    miu_n = n/mp.sqrt(2)
    sigma_n = mp.sqrt(n)/32**0.25
    z = NormalDist().inv_cdf(1-epsilon)
    k0 = mp.ceil(miu_n + (z*sigma_n))
    k1 = mp.floor(miu_n - z*sigma_n)
    ReZ = 0
    ImZ = 0
    k = 0
    while k <= k0:
        a = (-1)**k * C(k0, k1, miu_n, sigma_n, k) / ((k + 1)**sigma)
        b1 = t*mp.log(k+1)
        b2 = t*mp.log((k+1)/2)
        ReZ += a*(mp.cos(b1)-(2**(1-sigma))*mp.cos(b2))
        ImZ += a*(mp.sin(b1)-(2**(1-sigma))*mp.sin(b2))
        k+=1
    return mp.mpc(ReZ, -ImZ)/omega(sigma, t)

File: c++/zetaamtct.cpp
#include "zeta.hpp"
#include <functional>
#include <boost/math/tools/roots.hpp>
#include <iostream>
const static boost::math::normal distribution;

mpfr_float C(mpfr_float k0, mpfr_float k1, mpfr_float miu_n, mpfr_float sigma_n, int k)
{
    if (k0 < k1)
    {
        return 1;
    }
    return 1 - cdf(distribution, (k - miu_n)/sigma_n);
}

mpfr_float omega(mpfr_float sigma, mpfr_float t)
{
    return 1-(pow(2,2-sigma))*cos(t*log(2))+pow(4,1-sigma);
}
namespace zeta
{
    mpc_complex zetaAMTCT(mpc_complex s, int d)
    {
        mpfr_float sigma = s.real();
        mpfr_float t = s.imag();
        mpfr_float n = pi<mpfr_float>()* abs(t) 
            + (1 + 2*sigma)*log(abs(t)) 
            + 2*log(boost::math::tgamma(sigma)) 
            - 2 * log(abs(1 - pow(2,(1-sigma)))) 
            + 2 *d*log(10) 
            + log(2) 
            - log(pi<mpfr_float>());
        n = n/(2*log(3+sqrt(8)));
        mpfr_float epsilon = pow(10, -d);
        mpfr_float miu_n = n/sqrt(2);
        mpfr_float sigma_n = sqrt(n)/pow(32,0.25);
        mpfr_float z = boost::math::quantile(distribution,1-epsilon);
        mpfr_float k0 = ceil(miu_n + (z*sigma_n));
        mpfr_float k1 = floor(miu_n - (z*sigma_n));
        mpfr_float ReZ = 0;
        mpfr_float ImZ = 0;
        for (auto k = 0; k<= k0; k++)
        {
            mpfr_float a = pow(-1,k)*C(k0,k1,miu_n,sigma_n,k) / pow(k+1,sigma);
            mpfr_float b1 = t*log(k+1.);
            mpfr_float b2 = t*log((k+1)/2.);
            ReZ+= a*(cos(b1)-pow(2,1-sigma)*cos(b2));
            ImZ+= a*(sin(b1)-pow(2,1-sigma)*sin(b2));
        }
        return mpc_complex(ReZ, -ImZ)/omega(sigma, t);
    }
}